src/sidekick/csrf.js

```diff
@@ -0,0 +1,254 @@
+  log.warn('%j', {
+    csrf: {
+      org: org || '_top_level_',
+      site: org ? site || '_org_level_' : '',
+      origin: origin || '_missing_origin_',
+      extensionId,
+      message: `[CSRF Protection] ${message}`,
+      suffix,
+      referer,
+    },
+  });
+}
+
+/**
+ * Checks if a host is valid
+ * - only alphanumeric, dash, dot, and * for globbing
+ * - globbing is only allowed in the subdomains
+ * @param {string} host The host to check
+ * @returns {boolean} Whether the host is valid
+ */
+function checkHost(host) {
+  if (!host || typeof host !== 'string') {
+    return false;
+  }
+
+  if (host.length > 253) {
+    return false; // avoid regex DoS
+  }
+
+  // check for invalid characters and invalid character combinations
+  if (/[^a-zA-Z0-9\-.*]/.test(host) || host.includes('**') || host.includes('..')) {
+    return false;
+  }
+
+  // globbing is only allowed in the subdomains and maximum 2 globbing characters
+  if (host.includes('*')) {
+    if ([...host].filter((c) => c === '*').length > 2) {
+      return false;
+    }
+
+    const parts = host.split('.');
+    if (parts.length < 3 || parts[parts.length - 1].includes('*') || parts[parts.length - 2].includes('*')) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+/**
+ * Checks if a origin matches a host pattern that may include a glob
+ * @param {string} origin The origin to check
+ * @param {string} pattern The host pattern to match against
+ * @returns {boolean} Whether the host matches the pattern
+ */
+function hostMatches(origin, pattern) {
+  if (pattern.includes('*')) {
+    const regex = new RegExp(`^https://${pattern.replaceAll('.', '\\.').replaceAll('*', '[a-zA-Z0-9\\-\\.]{0,63}')}$`);
+    return regex.test(origin);
+  }
+  return origin === `https://${pattern}`;
+}
+
+/**
+ * Mitigates CSRF attacks through the sidekick token injection
+ * by validating that the request comes from a trusted origin from the browser.
+ * https://github.com/adobe/helix-admin/issues/2654
+ *
+ * @param {import('../support/AdminContext').AdminContext} context context
+ * @param {import('../support/RequestInfo').RequestInfo} info request info
+  *
+ * @throws {AccessDeniedError} if the request is not from a trusted origin
+ * @returns {Promise<void>}
+ */
+export async function sidekickCSRFProtection(context, info) {
+  if (!SIDEKICK_CSRF_PROTECTION_CONFIG.enabled) {
+    return;
+  }
+
+  const { headers = {} } = info;
+  const { log, attributes: { authInfo } } = context;
+
+  if (!authInfo || !authInfo.authenticated || !authInfo.extensionId) {
+    // CSRF only applies to authenticated requests
+    return;
+  }
+
+  const { extensionId } = authInfo;
+
+  const { org, site } = info;
+  const { suffix, method } = info;
+
+  // TODO: no longer required, `info` already contains correct org and site
+  // if (info.route === 'config') {
+  //   // special case for the config route
+  //   [, , org, , site] = info.suffix.split('.')[0].split(/\/+/);
+  // }
```

Comment:
I think this is no longer required, because `org` and `site` always come in the right place